### DISCUSSION TITLE
Does this mean that `mqtt` cannot be deployed in an Edge function? If so, why exactly?

### DISCUSSION THREAD

**Discussion by epatey** [vercel/next.js#58099] [App Router] [answered]
I have an API route that, among other things, publishes data to an `mqtt` topic when called. This works just fine in the `Node.js` runtime. When I tried to switch the route to the `Edge` runtime, though, it fails with a pretty obscure error message.

I've created a public [repo](https://github.com/epatey/mqttedgeproblem) with a minimal reproduction. The route is trivial. Here it is:
```typescript
import { connect } from "mqtt";
import { NextResponse } from "next/server";

export const runtime = "edge"; // Commenting this out makes it work

export function GET(): NextResponse<unknown> {
  connect("mqtt://test.mosquitto.org:1883"); // Commenting this out makes it work
  return NextResponse.json("hello");
}
```
and here's the error that occurs when I hit that route (it also fails when `pnpm build`ing).
```bash
 ✓ Compiled /route in 1416ms (236 modules)
 ⨯ node_modules/.pnpm/mqtt-packet@8.2.1/node_modules/mqtt-packet/parser.js (7:0) @ <unknown>
 ⨯ Class extends value [object Object] is not a constructor or null
```

To reproduce the problem:
- clone the repo
- `pnpm install`
- `pnpm dev`
- `curl localhost:3000` from another terminal

I assume this means that `mqtt` is somehow incompatible with the `Edge` runtime, but:
- I'd love to understand specifically what is incompatible
- It would be great if the error induced were more enlightening

Thanks for any thoughts on the subject.

**Comment by icyJoseph** [comment, marked as answer]
Most likely there's usage of Node APIs under the hood of `mqtt`. It can be several of these APIs, which means that the final error you see is obscure.

I took a quick look at `mqtt` and `mqtt-packet`, and the usage of Node APIs is heavy. 

`mqtt-packet` also uses `require`, which is not supported on edge environments. 

 - https://nextjs.org/docs/app/api-reference/edge#unsupported-apis

**Reply by epatey** [reply]
Thanks for the response, @icyJoseph. I'll go back and look at the `mqtt` code and try to see if I can identify usage of Node APIs. It makes perfect sense that they would be using them. Since I don't have experience using them myself, I didn't notice/recognize them.

**Reply by icyJoseph** [reply]
Keep in mind that it is Native Node APIs that are not supported on the Edge Runtime. Always check back with the list provided above.
